README
NAME
    Map::Tube::Text::Table - Table output for Map::Tube.

SYNOPSIS
     use Map::Tube::Text::Table;

     my $obj = Map::Tube::Text::Table->new(%params);
     my $text = $obj->junctions;
     my $text = $obj->line($line);
     my $text = $obj->lines;
     my $text = $obj->print;

METHODS
    "new(%params)"
             Constructor.

            *       "print_id"

                     Flag, that means printing of ID.
                     Affected methods:
                     - line()
                     - print() (by line()).
                     Default value is 0.

            *       "tube"

                     Map::Tube object.
                     It is required.
                     Default value is undef.

    "junctions()"
             Print junctions.
             Returns string with unicode text table.

    "line($line)"
             Print line.
             Returns string with unicode text table.

    "lines()"
             Print sorted lines.
             Returns string with unicode text table.

    "print()"
             Print all (junctions + all lines).
             Returns string with unicode text table.

ERRORS
     new():
             Parameter 'tube' is required.
             Parameter 'tube' must be 'Map::Tube' object.
             From Class::Utils::set_params():
                     Unknown parameter '%s'.

EXAMPLE
     use strict;
     use warnings;

     use Encode qw(encode_utf8);
     use English;
     use Error::Pure qw(err);
     use Map::Tube::Text::Table;

     # Error::Pure environment.
     $ENV{'ERROR_PURE'} = 'AllError';

     # Arguments.
     if (@ARGV < 1) {
             print STDERR "Usage: $0 metro\n";
             exit 1;
     }
     my $metro = $ARGV[0];
 
     # Object.
     my $class = 'Map::Tube::'.$metro;
     eval "require $class;";
     if ($EVAL_ERROR) {
             err "Cannot load '$class' class.",
                     'Error', $EVAL_ERROR;
     }
 
     # Metro object.
     my $tube = eval "$class->new";
     if ($EVAL_ERROR) {
             err "Cannot create object for '$class' class.",
                     'Error', $EVAL_ERROR;
     }
 
     # GraphViz object.
     my $table = Map::Tube::Text::Table->new(
             'tube' => $tube,
     );
 
     # Print out.
     print encode_utf8($table->print);

     # Output without arguments like:
     # Usage: /tmp/SZXfa2g154 metro

     # Output with 'Tbilisi' argument like:
     # ┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
     # │ Junctions                                                                                        │
     # ├──────────────────┬──────────────────────────────────────────┬────────────────────────────────────┤
     # │ Station          │ Line                                     │ Connected to                       │
     # ├──────────────────┼──────────────────────────────────────────┼────────────────────────────────────┤
     # │ სადგურის მოედანი │ ახმეტელი-ვარკეთილის ხაზი,საბურთალოს ხაზი │ მარჯანიშვილი, ნაძალადევი, წერეთელი │
     # └──────────────────┴──────────────────────────────────────────┴────────────────────────────────────┘
     # ┌───────────────────────────────────────────────────────────┐
     # │ Line 'ახმეტელი-ვარკეთილის ხაზი'                           │
     # ├──────────────────────┬────────────────────────────────────┤
     # │ Station              │ Connected to                       │
     # ├──────────────────────┼────────────────────────────────────┤
     # │ ახმეტელის თეატრი     │ სარაჯიშვილი                        │
     # │ სარაჯიშვილი          │ ახმეტელის თეატრი, გურამიშვილი      │
     # │ გურამიშვილი          │ სარაჯიშვილი, ღრმაღელე              │
     # │ ღრმაღელე             │ გურამიშვილი, დიდუბე                │
     # │ დიდუბე               │ გოცირიძე, ღრმაღელე                 │
     # │ გოცირიძე             │ დიდუბე, ნაძალადევი                 │
     # │ ნაძალადევი           │ გოცირიძე, სადგურის მოედანი         │
     # │ მარჯანიშვილი         │ რუსთაველი, სადგურის მოედანი        │
     # │ რუსთაველი            │ თავისუფლების მოედანი, მარჯანიშვილი │
     # │ თავისუფლების მოედანი │ ავლაბარი, რუსთაველი                │
     # │ ავლაბარი             │ 300 არაგველი, თავისუფლების მოედანი │
     # │ 300 არაგველი         │ ავლაბარი, ისანი                    │
     # │ ისანი                │ 300 არაგველი, სამგორი              │
     # │ სამგორი              │ ვარკეთილი, ისანი                   │
     # │ ვარკეთილი            │ სამგორი                            │
     # │ სადგურის მოედანი     │ მარჯანიშვილი, ნაძალადევი, წერეთელი │
     # └──────────────────────┴────────────────────────────────────┘
     # ┌────────────────────────────────────────────────────────────────────┐
     # │ Line 'საბურთალოს ხაზი'                                             │
     # ├─────────────────────────┬──────────────────────────────────────────┤
     # │ Station                 │ Connected to                             │
     # ├─────────────────────────┼──────────────────────────────────────────┤
     # │ წერეთელი                │ სადგურის მოედანი, ტექნიკური უნივერსიტეტი │
     # │ ტექნიკური უნივერსიტეტი  │ სამედიცინო უნივერსიტეტი, წერეთელი        │
     # │ სამედიცინო უნივერსიტეტი │ დელისი, ტექნიკური უნივერსიტეტი           │
     # │ დელისი                  │ ვაჟა-ფშაველა, სამედიცინო უნივერსიტეტი    │
     # │ ვაჟა-ფშაველა            │ დელისი                                   │
     # │ სადგურის მოედანი        │ მარჯანიშვილი, ნაძალადევი, წერეთელი       │
     # └─────────────────────────┴──────────────────────────────────────────┘

DEPENDENCIES
    Class::Utils, Error::Pure, Map::Tube::Text::Table::Utils,
    List::MoreUtils, Readonly, Scalar::Util.

SEE ALSO
    Task::Map::Tube
        Install the Map::Tube modules.

    Task::Map::Tube::Metro
        Install the Map::Tube concrete metro modules.

REPOSITORY
    <https://github.com/michal-josef-spacek/Map-Tube-Text-Table>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
     © 2014-2020 Michal Josef Špaček
     Artistic License
     BSD 2-Clause License

VERSION
    0.06



Changes
0.06

0.05 2020-01-19T13:27:16+01:00
 [Common]
 - Add author tests.
 - Add .travis.yml file.
 - Fix Makefile.PL for new version of Perl and no default »use lib '.'«.
 - Rewrite SEE ALSO section to new style.
 - Update Module::Install to 1.19 version.
 - Update author github username.
 - Update author name.
 - Update copyright years to 2020.

 [Map::Tube::Text::Table::Utils]
 - Add tests.
 - Improve SEE ALSO section in doc.

0.04 2015-01-24T18:42:18+01:00
 [Map::Tube::Text::Table]
 - Add Map::Tube::Graph to SEE ALSO section in doc.
 - Add Map::Tube::Plugin::Graph to SEE ALSO section in doc.
 - Add 'print_id' parameter for ID printing in line() method.
 - Fix junctions() method to new Map::Tube 2.72 version.
 - Fix line() method to new Map::Tube 2.72 version.

0.03 2015-01-07T07:53:38+01:00
 [Common]
 - Add Task::Map::Tube to SEE ALSO section in doc.
 - Update copyright years to 2015.

 [Map::Tube::Text::Table]
 - Add lines() method.
 - Rewrite to new Map::Tube::Text::Table::Utils::table().

 [Map::Tube::Text::Table::Utils]
 - Fix table() subroutine for situation without headers.
 - Fix indentation in table() subroutine.

0.02 2014-12-19T19:33:51+01:00
 - Fix dependencies.
 - Update META.yml file.

0.01 2014-12-19T08:21:12+01:00
 - First version.


Makefile.PL
use lib 'inc';
use strict;
use warnings;

# Modules.
use inc::Module::Install;

# Definition.
abstract 'Table information about Map::Tube object.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.08;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
license 'artistic_2';
name 'Map-Tube-Text-Table';
readme_from 'Table.pm';
requires 'Class::Utils' => 0.06;
requires 'Error::Pure' => 0.21;
requires 'List::Util' => 0;
requires 'List::MoreUtils' => 0;
requires 'Readonly' => 0;
requires 'Scalar::Util' => 0;
requires 'Text::UnicodeBox' => 0;
requires 'Text::UnicodeBox::Control' => 0;
requires 'perl' => '5.8.0';
resources 'bugtracker' => 'https://rt.cpan.org/Public/Dist/Display.html'.
	'?Name=Map-Tube-Text-Table';
resources 'homepage' => 'https://github.com/michal-josef-spacek/Map-Tube-Text-Table';
resources 'repository' => 'git://github.com/michal-josef-spacek/Map-Tube-Text-Table';
test_requires 'English' => 0;
test_requires 'File::Object' => 0.08;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
tests_recursive;
version '0.06';

# Run.
WriteAll();


Table/Utils.pm
package Map::Tube::Text::Table::Utils;

use base qw(Exporter);
use strict;
use warnings;

use List::Util qw(sum);
use Readonly;
use Text::UnicodeBox;
use Text::UnicodeBox::Control qw(:all);

# Constants.
Readonly::Array our @EXPORT_OK => qw(table);
Readonly::Scalar our $EMPTY_STR => q{};
Readonly::Scalar our $SPACE => q{ };
Readonly::Scalar our $SPACE_ON_END_COUNT => 1;

our $VERSION = 0.06;

# Print table.
sub table {
	my ($title, $data_len_ar, $header_ar, $data_ar) = @_;

	# Check data.
	if (! @{$data_ar}) {
		return $EMPTY_STR;
	}

	my $t = Text::UnicodeBox->new;

	# Table title.
	my $pipes_in_count = @{$data_len_ar} * 2 - 2;
	$t->add_line(
		BOX_START('bottom' => 'light', 'top' => 'light'),
		_column_left($title, sum(map { $_ + $SPACE_ON_END_COUNT }
			@{$data_len_ar}) + $pipes_in_count),
		BOX_END(),
	);

	# Legend.
	if (defined $header_ar) {
		$t->add_line(
			BOX_START('bottom' => 'light', 'top' => 'light'),
			_columns($header_ar, $data_len_ar),
		);
	}

	# Data.
	while (my $row_ar = shift @{$data_ar}) {
		$t->add_line(
			BOX_START(
				@{$data_ar} == 0 ? ('bottom' => 'light') : (),
			),
			_columns($row_ar, $data_len_ar),
		);
	}

	# Render to output.
	return $t->render;
}

# Column text with left align.
sub _column_left {
	my ($text, $width) = @_;
	my $text_len = length $text;
	return $SPACE.$text.($SPACE x ($width - $text_len));
}

# Get Text::UnicodeBox columns.
sub _columns {
	my ($data_ar, $data_len_ar) = @_;
	my @ret;
	my $i = 0;
	foreach my $item (@{$data_ar}) {
		push @ret, _column_left($item, $data_len_ar->[$i++]
			+ $SPACE_ON_END_COUNT);
		if (@{$data_ar} > $i) {
			push @ret, BOX_RULE;
		} else {
			push @ret, BOX_END;
		}
	}
	return @ret;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Map::Tube::Text::Table::Utils - Utilities for Map::Tube::Text::Table.

=head1 SYNOPSIS

 use Map::Tube::Text::Table::Utils qw(table);

 my $table = table($title, $data_len_ar, $header_ar, $data_ar);

=head1 SUBROUTINES

=over 8

=item C<table($title, $data_len_ar, $header_ar, $data_ar)>

 Print table.
 Returns text.

=back

=head1 EXAMPLE1

 use strict;
 use warnings;

 use Encode qw(encode_utf8);
 use Map::Tube::Text::Table::Utils qw(table);

 # Get table.
 my $table = table('Title', [1, 2, 3], ['A', 'BB', 'CCC'], [
         ['E', 'A', 'A'],
         ['A', 'Ga', 'Acv'],
 ]);

 # Print table.
 print encode_utf8($table);

 # Output:
 # ┌──────────────┐
 # │ Title        │
 # ├───┬────┬─────┤
 # │ A │ BB │ CCC │
 # ├───┼────┼─────┤
 # │ E │ A  │ A   │
 # │ A │ Ga │ Acv │
 # └───┴────┴─────┘

=head1 DEPENDENCIES

L<Exporter>,
L<List::Util>,
L<Readonly>,
L<Text::UnicodeBox>,
L<Text::UnicodeBox::Control>.

=head1 SEE ALSO

=over 8

=item L<Map::Tube>

Lightweight Routing Framework

=item L<Map::Tube::Text::Table>

Table output for Map::Tube

=item L<Task::Map::Tube>

Install the Map::Tube modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Map-Tube-Text-Table>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

 © 2014-2020 Michal Josef Špaček
 Artistic License
 BSD 2-Clause License

=head1 VERSION

0.06

=cut


Table.pm
package Map::Tube::Text::Table;

use strict;
use warnings;

use Class::Utils qw(set_params);
use Error::Pure qw(err);
use Map::Tube::Text::Table::Utils qw(table);
use List::MoreUtils qw(any);
use Readonly;
use Scalar::Util qw(blessed);

# Constants.
Readonly::Scalar our $CONNECTED_TO => q{Connected to};
Readonly::Scalar our $ID => q{ID};
Readonly::Scalar our $JUNCTIONS => q{Junctions};
Readonly::Scalar our $LINE => q{Line};
Readonly::Scalar our $LINES => q{Lines};
Readonly::Scalar our $STATION => q{Station};

our $VERSION = 0.06;

# Constructor.
sub new {
	my ($class, @params) = @_;

	# Create object.
	my $self = bless {}, $class;

	# Print ids.
	$self->{'print_id'} = 0;

	# Map::Tube object.
	$self->{'tube'} = undef;

	# Process params.
	set_params($self, @params);

	# Check Map::Tube object.
	if (! defined $self->{'tube'}) {
		err "Parameter 'tube' is required.";
	}
	if (! blessed($self->{'tube'})
		|| ! $self->{'tube'}->does('Map::Tube')) {

		err "Parameter 'tube' must be 'Map::Tube' object.";
	}

	# Object.
	return $self;
}

# Print junctions.
sub junctions {
	my $self = shift;

	# Get data.
	my @data;
	my @title = ($STATION, $LINE, $CONNECTED_TO);
	my @data_len = map { length $_ } @title;
	my $nodes_hr = $self->{'tube'}->nodes;
	foreach my $node_name (sort keys %{$nodes_hr}) {
		if (@{$nodes_hr->{$node_name}->line} > 1) {

			# Get data.
			my @links = map { $self->{'tube'}->get_node_by_id($_)->name }
				split m/,/ms, $nodes_hr->{$node_name}->link;
			my $data_ar = [
				$nodes_hr->{$node_name}->name,
				(join ', ', map { $_->name } @{$nodes_hr->{$node_name}->line}),
				(join ', ', sort @links),
			];
			push @data, $data_ar;

			# Maximum data length.
			foreach my $i (0 .. $#{$data_ar}) {
				if (length $data_ar->[$i] > $data_len[$i]) {
					$data_len[$i] = length $data_ar->[$i];
				}
			}
		}
	}

	# Print and return table.
	return table($JUNCTIONS, \@data_len, \@title, \@data);
}

# Print line.
sub line {
	my ($self, $line) = @_;

	# Get data.
	my @data;
	my @title = ($STATION, $CONNECTED_TO);
	if ($self->{'print_id'}) {
		unshift @title, $ID;
	}
	my @data_len = map { length $_ } @title;
	my $nodes_hr = $self->{'tube'}->nodes;
	foreach my $node_name (sort keys %{$nodes_hr}) {
		if (any { $_ eq $line } map { $_->name }
			@{$nodes_hr->{$node_name}->line}) {

			# Get data.
			my @links = map { $self->{'tube'}->get_node_by_id($_)->name }
				split m/,/ms, $nodes_hr->{$node_name}->link;
			my $data_ar = [
				$nodes_hr->{$node_name}->name,
				(join ', ', sort @links),
			];
			if ($self->{'print_id'}) {
				unshift @{$data_ar}, $nodes_hr->{$node_name}->id,
			}
			push @data, $data_ar;

			# Maximum data length.
			foreach my $i (0 .. $#{$data_ar}) {
				if (length $data_ar->[$i] > $data_len[$i]) {
					$data_len[$i] = length $data_ar->[$i];
				}
			}
		}
	}

	# Print and return table.
	return table($LINE." '$line'", \@data_len, \@title, \@data);
}

# Get lines.
sub lines {
	my $self = shift;
	my $lines_ar = $self->{'tube'}->get_lines;
	my $length = 0;
	my @data;
	foreach my $line (sort @{$lines_ar}) {
		push @data, [$line];
		if (length $line > $length) {
			$length = length $line;
		}
	}
	return table($LINES, [$length], undef, \@data);
}

# Print all.
sub print {
	my $self = shift;
	my $ret = $self->junctions;
	foreach my $line (@{$self->{'tube'}->get_lines}) {
		$ret .= $self->line($line->name);
	}
	return $ret;
}

1;

__END__

=encoding utf8

=head1 NAME

Map::Tube::Text::Table - Table output for Map::Tube.

=head1 SYNOPSIS

 use Map::Tube::Text::Table;

 my $obj = Map::Tube::Text::Table->new(%params);
 my $text = $obj->junctions;
 my $text = $obj->line($line);
 my $text = $obj->lines;
 my $text = $obj->print;

=head1 METHODS

=over 8

=item C<new(%params)>

 Constructor.

=over 8

=item * C<print_id>

 Flag, that means printing of ID.
 Affected methods:
 - line()
 - print() (by line()).
 Default value is 0.

=item * C<tube>

 Map::Tube object.
 It is required.
 Default value is undef.

=back

=item C<junctions()>

 Print junctions.
 Returns string with unicode text table.

=item C<line($line)>

 Print line.
 Returns string with unicode text table.

=item C<lines()>

 Print sorted lines.
 Returns string with unicode text table.

=item C<print()>

 Print all (junctions + all lines).
 Returns string with unicode text table.

=back

=head1 ERRORS

 new():
         Parameter 'tube' is required.
         Parameter 'tube' must be 'Map::Tube' object.
         From Class::Utils::set_params():
                 Unknown parameter '%s'.

=head1 EXAMPLE

 use strict;
 use warnings;

 use Encode qw(encode_utf8);
 use English;
 use Error::Pure qw(err);
 use Map::Tube::Text::Table;

 # Error::Pure environment.
 $ENV{'ERROR_PURE'} = 'AllError';

 # Arguments.
 if (@ARGV < 1) {
         print STDERR "Usage: $0 metro\n";
         exit 1;
 }
 my $metro = $ARGV[0];
 
 # Object.
 my $class = 'Map::Tube::'.$metro;
 eval "require $class;";
 if ($EVAL_ERROR) {
         err "Cannot load '$class' class.",
                 'Error', $EVAL_ERROR;
 }
 
 # Metro object.
 my $tube = eval "$class->new";
 if ($EVAL_ERROR) {
         err "Cannot create object for '$class' class.",
                 'Error', $EVAL_ERROR;
 }
 
 # GraphViz object.
 my $table = Map::Tube::Text::Table->new(
         'tube' => $tube,
 );
 
 # Print out.
 print encode_utf8($table->print);

 # Output without arguments like:
 # Usage: /tmp/SZXfa2g154 metro

 # Output with 'Tbilisi' argument like:
 # ┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
 # │ Junctions                                                                                        │
 # ├──────────────────┬──────────────────────────────────────────┬────────────────────────────────────┤
 # │ Station          │ Line                                     │ Connected to                       │
 # ├──────────────────┼──────────────────────────────────────────┼────────────────────────────────────┤
 # │ სადგურის მოედანი │ ახმეტელი-ვარკეთილის ხაზი,საბურთალოს ხაზი │ მარჯანიშვილი, ნაძალადევი, წერეთელი │
 # └──────────────────┴──────────────────────────────────────────┴────────────────────────────────────┘
 # ┌───────────────────────────────────────────────────────────┐
 # │ Line 'ახმეტელი-ვარკეთილის ხაზი'                           │
 # ├──────────────────────┬────────────────────────────────────┤
 # │ Station              │ Connected to                       │
 # ├──────────────────────┼────────────────────────────────────┤
 # │ ახმეტელის თეატრი     │ სარაჯიშვილი                        │
 # │ სარაჯიშვილი          │ ახმეტელის თეატრი, გურამიშვილი      │
 # │ გურამიშვილი          │ სარაჯიშვილი, ღრმაღელე              │
 # │ ღრმაღელე             │ გურამიშვილი, დიდუბე                │
 # │ დიდუბე               │ გოცირიძე, ღრმაღელე                 │
 # │ გოცირიძე             │ დიდუბე, ნაძალადევი                 │
 # │ ნაძალადევი           │ გოცირიძე, სადგურის მოედანი         │
 # │ მარჯანიშვილი         │ რუსთაველი, სადგურის მოედანი        │
 # │ რუსთაველი            │ თავისუფლების მოედანი, მარჯანიშვილი │
 # │ თავისუფლების მოედანი │ ავლაბარი, რუსთაველი                │
 # │ ავლაბარი             │ 300 არაგველი, თავისუფლების მოედანი │
 # │ 300 არაგველი         │ ავლაბარი, ისანი                    │
 # │ ისანი                │ 300 არაგველი, სამგორი              │
 # │ სამგორი              │ ვარკეთილი, ისანი                   │
 # │ ვარკეთილი            │ სამგორი                            │
 # │ სადგურის მოედანი     │ მარჯანიშვილი, ნაძალადევი, წერეთელი │
 # └──────────────────────┴────────────────────────────────────┘
 # ┌────────────────────────────────────────────────────────────────────┐
 # │ Line 'საბურთალოს ხაზი'                                             │
 # ├─────────────────────────┬──────────────────────────────────────────┤
 # │ Station                 │ Connected to                             │
 # ├─────────────────────────┼──────────────────────────────────────────┤
 # │ წერეთელი                │ სადგურის მოედანი, ტექნიკური უნივერსიტეტი │
 # │ ტექნიკური უნივერსიტეტი  │ სამედიცინო უნივერსიტეტი, წერეთელი        │
 # │ სამედიცინო უნივერსიტეტი │ დელისი, ტექნიკური უნივერსიტეტი           │
 # │ დელისი                  │ ვაჟა-ფშაველა, სამედიცინო უნივერსიტეტი    │
 # │ ვაჟა-ფშაველა            │ დელისი                                   │
 # │ სადგურის მოედანი        │ მარჯანიშვილი, ნაძალადევი, წერეთელი       │
 # └─────────────────────────┴──────────────────────────────────────────┘

=head1 DEPENDENCIES

L<Class::Utils>,
L<Error::Pure>,
L<Map::Tube::Text::Table::Utils>,
L<List::MoreUtils>,
L<Readonly>,
L<Scalar::Util>.

=head1 SEE ALSO

=over

=item L<Task::Map::Tube>

Install the Map::Tube modules.

=item L<Task::Map::Tube::Metro>

Install the Map::Tube concrete metro modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Map-Tube-Text-Table>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

 © 2014-2020 Michal Josef Špaček
 Artistic License
 BSD 2-Clause License

=head1 VERSION

0.06

=cut
